Refactor GCEClient: wrap compute.Service in an interface for mocking …

spew opened the pull request:
…GCP compute

This change creates a ComputeService implementation which has a runtime
implementation that wraps the compute.Service. The MachineActuator is
changed to make use of the ComputeService through a new interface named
GCEClientComputeService. This will enable creating tests that mock GCP Compute
Service calls to control MachineActuator behavior.

**What this PR does / why we need it**:
It is difficult to test changes that interact with GCP without manually running integration tests. This change will enable developers to write unit tests that mock GCP and test that their code correctly handles various scenarios.

**Release note**:
```release-note
NONE
```

<!-- All reviews default to cc'ing the kube-deploy-reviewers github group. -->
@kubernetes/kube-deploy-reviewers

krousey commented:
/test all

spew commented:
/retest

krousey commented:
/retest

krousey commented on cloud/google/clients/computeservice_test.go:
```diff
@@ -0,0 +1,121 @@
+	}
+	mux.Handle("/compute/v1/projects/projectName/global/images/imageName", handler(nil, &responseImage))
+	image, err := client.ImagesGet("projectName", "imageName")
+	assert.Nil(t, err)
```
If you're going to use and assertion library, at least provide useful failure messages. This is hardly helpful:

```go
func TestFoo(t *testing.T) {                                                                                                                                                                              
    x := 1                                                                                                                                                                                                
    y := 2                                                                                                                                                                                                
    assert.Equal(t, x, y)                                                                                                                                                                                        
} 
```

```console
go test                                                                                                                                                                                                   
--- FAIL: TestFoo (0.00s)                                                                                                                                                                                 
        Error Trace:                                                                                                                                                                                      
        Error:          Not equal: 1 (expected)                                                                                                                                                           
                                != 2 (actual)
```

Now for a more opinionated comment, please don't use assertion libraries if you can. This a request, not a mandate. I would state my reasons, but the Go team sums them up pretty neatly: https://golang.org/doc/faq#testing_framework.

spew replied:
I think the go testing library does support letting the test finish and then all failed assertions will print. I sort of understand the anti-assertion library sentiment. I think one good thing about using the library is that the error messages are consistent and easy to scan & read. However, as you pointed out, that's not very useful if the failure messages aren't populated in which case there is a bit of a custom message anyway.

I was looking through some of the k8s.io libraries and it seems that there is a mix of using the assert library and implementing the checks. I'll go with what you suggested and rework these tests to not u se assert since it seems that is somewhat the norm.

spew commented:
Updated with the following changes:
1. Removed the 'newComputeService' method and moved the related code into NewComputeService
2. Removed usage of the assertion library in tests and simply check the various scenarios.

spew commented:
Been out -- rebased on top of the latest commits so this can be reviewed again.

spew commented on cloud/google/clients/computeservice.go:
```diff
@@ -0,0 +1,58 @@
+	"net/url"
+)
+
+type ComputeService struct {
```
I've updated with some minimal documentation that I thought was appropriate, happy to change or add more if you have something different in mind.

krousey commented:
/lgtm

krousey commented:
/approve
